### PR TITLE
[tune] Implement median stopping rule

### PR DESCRIPTION
**Pull request by ericl**
This is the median stopping rule as described in the Vizier paper: https://research.google.com/pubs/pub46180.html

See the docstring of `MedianStoppingRule` for usage. I plan to expose this in the command line API once the variant generation changes land to avoid merge conflicts.

### CODE REVIEW

File: python/ray/tune/trial_scheduler.py [hunk 1]
```diff
@@ -44,3 +59,72 @@ def choose_trial_to_run(self, trial_runner):
...
+# TODO(ekl) expose this in the command line API
+class MedianStoppingRule(FIFOScheduler):
+    """Implements the median stopping rule as described in the Vizier paper:
+
+        https://research.google.com/pubs/pub46180.html
+
+    Args:
+        time_attr (str): The TrainingResult attr to use for comparing time.
+        reward_attr (str): The TrainingResult objective value attribute.
+        grace_period (float): Only stop trials at least this old in time.
```
Comment (richardliaw) [inline python/ray/tune/trial_scheduler.py]: Would be nice to remind units are in sec.

File: python/ray/tune/trial_scheduler.py [hunk 2]
```diff
@@ -44,3 +59,72 @@ def choose_trial_to_run(self, trial_runner):
...
+
+
+# TODO(ekl) expose this in the command line API
+class MedianStoppingRule(FIFOScheduler):
+    """Implements the median stopping rule as described in the Vizier paper:
+
+        https://research.google.com/pubs/pub46180.html
+
+    Args:
+        time_attr (str): The TrainingResult attr to use for comparing time.
```
Comment (richardliaw) [inline python/ray/tune/trial_scheduler.py]: Could imagine that one would want to use want to use `iteration` or something non-temporal as a metric of progress.

File: python/ray/tune/trial_scheduler.py [hunk 3]
```diff
@@ -44,3 +59,72 @@ def choose_trial_to_run(self, trial_runner):
...
+        FIFOScheduler.__init__(self)
+        self._completed_trials = set()
+        self._results = collections.defaultdict(list)
+        self._grace_period = grace_period
+        self._min_samples_required = min_samples_required
+        self._reward_attr = reward_attr
+        self._time_attr = time_attr
+        self._num_stopped = 0
+
+    def on_trial_result(self, trial_runner, trial, result):
```
Comment (richardliaw) [inline python/ray/tune/trial_scheduler.py]: Would be nice to add some comments - 
```
"""
This stopping rule stops a running trial if the trial’s best objective
value by step `t` is strictly worse than the median value of the running 
averages of all completed trials’ objectives reported up to step `t`.
"""
```

File: test/trial_scheduler_test.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+from ray.tune.result import TrainingResult
+from ray.tune.trial import Trial
+from ray.tune.trial_scheduler import MedianStoppingRule, TrialScheduler
+
+
+def result(t, rew):
+    return TrainingResult(time_total_s=t, episode_reward_mean=rew)
+
+
+class EarlyStoppingSuite(unittest.TestCase):
```
Comment (richardliaw) [inline test/trial_scheduler_test.py]: Can we also cover `time_attr='time_total_s'`, `reward_attr='episode_reward_mean'` initializer parameters?

**Review by richardliaw — COMMENTED** [review]
Mainly looks good, can approve after comments

File: python/ray/tune/trial_scheduler.py (hunk 2 above)
Comment (ericl) [reply python/ray/tune/trial_scheduler.py]: Added comment.

File: python/ray/tune/trial_scheduler.py (hunk 1 above)
Comment (ericl) [reply python/ray/tune/trial_scheduler.py]: Added comment.

File: python/ray/tune/trial_scheduler.py (hunk 3 above)
Comment (ericl) [reply python/ray/tune/trial_scheduler.py]: Added comment.

File: test/trial_scheduler_test.py (hunk 4 above)
Comment (ericl) [reply test/trial_scheduler_test.py]: Done

**Review by richardliaw — APPROVED** [review]

**Pull request merged**